lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,339 @@
+                        overallField,
+                        hydration,
+                        argument = actorArg.name,
+                    )
+                } else {
+                    null
+                }
+            }
+
+        return duplicatedArgumentsErrors + remoteArgErrors + missingActorArgErrors
+    }
+
+    private fun getRemoteArgErrors(
+        parent: NadelServiceSchemaElement,
+        overallField: GraphQLFieldDefinition,
+        remoteArg: RemoteArgumentDefinition,
+        actorField: GraphQLFieldDefinition,
+    ): List<NadelSchemaValidationError> {
+        val remoteArgSource = remoteArg.remoteArgumentSource
+        val actorArg = actorField.getArgument(remoteArg.name)
+        val actorArgInputType = actorArg.type
+        return when (remoteArgSource.sourceType) {
+            ObjectField -> {
+                val field = (parent.underlying as GraphQLFieldsContainer).getFieldAt(remoteArgSource.pathToField!!)
+                if (field == null) {
+                    listOf(
+                        MissingHydrationFieldValueSource(parent, overallField, remoteArgSource)
+                    )
+                } else {
+                    //check the input types match with hydration and actor fields
+                    val fieldOutputType = field.type
+                    return hydrationArgTypesMatch(parent, overallField, remoteArg, fieldOutputType, actorArgInputType)
+                }
+            }
+            FieldArgument -> {
+                val argument = overallField.getArgument(remoteArgSource.argumentName!!)
+                if (argument == null) {
+                    listOf(MissingHydrationArgumentValueSource(parent, overallField, remoteArgSource))
+                } else {
+                    //check the input types match with hydration and actor fields
+                    val hydrationArgType = argument.type
+                    return hydrationArgTypesMatch(parent, overallField, remoteArg, hydrationArgType, actorArgInputType)
+                }
+            }
+            else -> {
+                emptyList()
+            }
+        }
+    }
+
+    /*
+    *  Checks the type of a hydration argument derived from either a $source field output type or $argument field input
+    * argument type against the type of an actor field argument to see if they match
+    * will also allow for cases of batched hydration eg ID! -> [ID]
+    */
+    private fun hydrationArgTypesMatch(
+        parent: NadelServiceSchemaElement,
+        overallField: GraphQLFieldDefinition,
+        remoteArg: RemoteArgumentDefinition,
+        hydrationSourceType: GraphQLType,
+        actorArgumentInputType: GraphQLType,
+    ): List<NadelSchemaValidationError> {
+
+        var isIncompatible = false
+
+        var hydrationSource: GraphQLType = hydrationSourceType
+        var actorArgument: GraphQLType = actorArgumentInputType
+
+        while (hydrationSource.isWrapped && actorArgument.isWrapped) {
```

Comment:
I'm really not sure about this code… Can you explain it more? Perhaps it's okay.

I really think we should have checked whether the hydration was batched or not before taking any action… This code makes too many assumptions on whether something is batched or not and there's a lot of paths to follow.

Reply:
yea so I found over here was really the only appropriate place I can check to see if it's a batch hydration, as I need to unravel and check the actor argument input type and the hydration source type, to see if one is a list and the other is a single value, there's no where before this execution that could indicate whether a hydration is batched or not, even the `isBatched` function checks if the `actorFieldDef` is a list type